### DISCUSSION TITLE
How to use OpenMP the right way?

### DISCUSSION THREAD

**Discussion by stefanos82** [edubart/nelua-lang#245] [Q&A] [answered]
I have an example code you shared somewhere in here a couple of years ago (...already years?!) and I'm trying to take advantage of OpenMP parallelism's capabilities.

Here's my attempt:

```lua
##[[
if ccinfo.is_gcc and not ccinfo.is_clang then
  OPENMP = true
  cflags '-DLINUX -D_REENTRANT -D_GNU_SOURCE -I/usr/include/apr-1.0 -fopenmp'
  ldflags '-lapr-1'
  cinclude '<apr_pools.h>'
  linklib 'gomp'
end
]]
local function fibmod(n: uinteger, m: uinteger)
  local a, b = 0_u, 1_u

##[[if OPENMP then
    cemit '#pragma omp parallel for schedule(dynamic)'
end]]

  for i=1_u,n do
    a, b = b, (a + b) % m
  end
  return a
end

local res = fibmod(100000000, 1000000000000)
assert(res == 167760546875)
print(res)
```

The error I'm getting is the following:

```c
/home/stefanos/.cache/nelua/fibmod.c: In function ‘fibmod_fibmod’:
/home/stefanos/.cache/nelua/fibmod.c:141:34: error: expected iteration declaration or initialization before ‘i’
  141 |   for(uint64_t i = 1U, _end = n; i <= _end; i += 1) {
      |                                  ^
error: C compilation for '/home/stefanos/.cache/nelua/fibmod' failed

shell returned 1
```

@edubart what is the right way I need to configure my pragma in order to make it work?

**Comment by jrfondren** [comment, marked as answer]
Each iteration of your loop depends on the previous iteration's values, so I don't see how it can be parallelized. Maybe the error is OpenMP's way of complaining about that.

Here's a recursive fibonacci and a use of it that's speed up with OpenMP. The usage is the same as in examples/condots.nelua:

```lua
local function fib(n: integer)
  if n < 2 then return n end
  return fib(n - 1) + fib(n - 2)
end

local results: [1000]uinteger
## if pragmas.withmp then
  ## cflags '-fopenmp'
  ## cemit '#pragma omp parallel for schedule(dynamic)'
## end
for i=0, #results-1 do
  results[i] = fib(30)
end
print(results[0])
```

comparing them:
```
$ nelua -o fib1 -M fibopenmp.nelua
$ nelua -o fib2 -P withmp -M fibopenmp.nelua
$ hyperfine ./fib1 ./fib2
Benchmark 1: ./fib1
  Time (mean ± σ):      1.185 s ±  0.009 s    [User: 1.182 s, System: 0.000 s]
  Range (min … max):    1.170 s …  1.196 s    10 runs
 
Benchmark 2: ./fib2
  Time (mean ± σ):     162.6 ms ±   3.5 ms    [User: 2529.6 ms, System: 3.6 ms]
  Range (min … max):   157.8 ms … 172.7 ms    18 runs
 
Summary
  ./fib2 ran
    7.29 ± 0.17 times faster than ./fib1
```

**Reply by stefanos82** [reply]
This is the type of programming I'm talking about, amazing stuff!

Thank you @jrfondren :+1:

**Comment by jrfondren** [comment]
Belatedly, OpenMP's actual problem with your code is that it doesn't accept multiple initialization in a loop, and nelua easily emits that. I only chanced upon a case with `#somearray` where it doesn't.

**Reply by stefanos82** [reply]
Yeah, I verified that with `--print-code` after your initial response and saw that `for i=0, #results-1 do` is being emitted as `for(int64_t i = 0; i <= 999; i += 1) {` ...

Using OpenMP is quite complicated, I have to admit; I have downloaded their reference manual and have found examples that look quite sophisticated to my liking.
